### PR TITLE
feat(ingest): Glue jobs

### PR DESCRIPTION
**Pull request by kevinhu**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: gms/api/src/main/snapshot/com.linkedin.datajob.dataJobs.snapshot.json [hunk 1]
```diff
@@ -438,9 +438,10 @@
             "name" : "AzkabanJobType",
             "namespace" : "com.linkedin.datajob.azkaban",
             "doc" : "The various types of support azkaban jobs",
-            "symbols" : [ "COMMAND", "HADOOP_JAVA", "HADOOP_SHELL", "HIVE", "PIG", "SQL" ],
```
Comment (hsheth2) [inline gms/api/src/main/snapshot/com.linkedin.datajob.dataJobs.snapshot.json:441]: Let's not keep piling onto the `AzkabanJobType` 

cc @jjoyce0510 is the plan still to add a free form string?

File: datahub-web-react/src/utils/sort/topologicalSort.ts [hunk 2]
```diff
@@ -16,11 +17,14 @@ function topologicalSortHelper(
         .filter((entity) => entity?.entity?.urn && urnsArray.includes(entity?.entity?.urn))
         .forEach((n) => {
             if (n?.entity?.urn && !explored.has(n?.entity?.urn)) {
-                topologicalSortHelper(n, explored, result, urnsArray);
+                topologicalSortHelper(n, explored, result, urnsArray, nodes);
             }
         });
     if (urnsArray.includes(node?.entity?.urn)) {
-        result.push(node);
+        const fullyFetchedEntity = nodes.find((n) => n?.entity?.urn === node?.entity?.urn);
```
Comment (jjoyce0510) [inline datahub-web-react/src/utils/sort/topologicalSort.ts:24]: What's the rationale for this change?

File: gms/api/src/main/snapshot/com.linkedin.datajob.dataJobs.snapshot.json (hunk 1 above)
Comment (jjoyce0510) [reply gms/api/src/main/snapshot/com.linkedin.datajob.dataJobs.snapshot.json:441]: Yeah. This can either be a freeform string or another enum with a better name. 

I don't have a strong preference for adding a freeform string vs a better enum

File: datahub-web-react/src/utils/sort/topologicalSort.ts (hunk 2 above)
Comment (kevinhu) [reply datahub-web-react/src/utils/sort/topologicalSort.ts:24]: The previous implementation was causing a bug where nodes only had urns and types specified as they were from upstreamLineage – @gabe-lyons can elaborate!

File: metadata-ingestion/src/datahub/ingestion/source/glue.py [hunk 3]
```diff
@@ -131,44 +151,270 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
...
+        url = urlparse(script_path, allow_fragments=False)
+        bucket = url.netloc
+        key = url.path[1:]
+
+        # download the script contents
+        obj = self.s3_client.get_object(Bucket=bucket, Key=key)
+        script = obj["Body"].read().decode("utf-8")
+
+        # extract the job DAG from the script
+        return self.glue_client.get_dataflow_graph(PythonScript=script)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/glue.py:196]: interesting. didn't realize you are getting the graph from the script itself..

File: metadata-ingestion/src/datahub/ingestion/source/glue.py [hunk 4]
```diff
@@ -131,44 +151,270 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
...
+                        f"{node_args['database']}.{node_args['table_name']}"
+                    )
+
+                    # we know that the table will already be covered when ingesting Glue tables
+                    node_urn = f"urn:li:dataset:(urn:li:dataPlatform:glue,{full_table_name},{self.env})"
+
+                # if data object is S3 bucket
+                elif node_args.get("connection_type") == "s3":
+
+                    # remove S3 prefix
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/glue.py:238]: may be worthwhile to comment what prefix you're removing... eg. s3:// :)

File: metadata-ingestion/src/datahub/ingestion/source/glue.py [hunk 5]
```diff
@@ -131,44 +151,270 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
...
+                    )
+
+                    dataset_snapshot = DatasetSnapshot(
+                        urn=node_urn,
+                        aspects=[],
+                    )
+
+                    dataset_snapshot.aspects.append(Status(removed=False))
+                    dataset_snapshot.aspects.append(
+                        OwnershipClass(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/glue.py:255]: Would be great if there were a way to resolve owners from the AWS IAM roles associated with this table.. 

Or the schema of the S3 file. I guess we'd need something to parse the extension / file to do that?

File: metadata-ingestion/src/datahub/ingestion/source/glue.py [hunk 6]
```diff
@@ -131,44 +151,270 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
...
+            return {
+                **node,
+                "urn": node_urn,
+                # to be filled in after traversing edges
+                "inputDatajobs": [],
+                "inputDatasets": [],
+                "outputDatasets": [],
+            }
+
+        nodes = {node["Id"]: process_node(node) for node in dataflow_graph["DagNodes"]}
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/glue.py:294]: Maybe its just me, but the nested function makes this slightly difficult to follow

**Review by jjoyce0510 — APPROVED** [review]
Overall this is looking good to me

File: metadata-ingestion/src/datahub/ingestion/source/glue.py (hunk 3 above)
Comment (kevinhu) [reply metadata-ingestion/src/datahub/ingestion/source/glue.py:196]: Yeah, I thought they would have a method for getting it directly, but seems like this is the only way to get it through boto

File: metadata-ingestion/src/datahub/ingestion/source/glue.py (hunk 6 above)
Comment (kevinhu) [reply metadata-ingestion/src/datahub/ingestion/source/glue.py]: Fixed!

File: metadata-ingestion/src/datahub/ingestion/source/glue.py (hunk 4 above)
Comment (kevinhu) [reply metadata-ingestion/src/datahub/ingestion/source/glue.py]: Commented!

File: metadata-ingestion/src/datahub/ingestion/source/glue.py [hunk 7]
```diff
@@ -131,44 +151,275 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
...
+        # download the script contents
+        # see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_object
+        obj = self.s3_client.get_object(Bucket=bucket, Key=key)
+        script = obj["Body"].read().decode("utf-8")
+
+        # extract the job DAG from the script
+        # see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/glue.html#Glue.Client.get_dataflow_graph
+        return self.glue_client.get_dataflow_graph(PythonScript=script)
+
+    def process_dataflow_graph(self, dataflow_graph, flow_urn):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/glue.py:201]: can we add some type annotations here

**Review by hsheth2 — COMMENTED** [review]
LGTM with one nit

File: metadata-ingestion/src/datahub/ingestion/source/glue.py (hunk 5 above)
Comment (kevinhu) [reply metadata-ingestion/src/datahub/ingestion/source/glue.py:260]: Yeah so this is already done for owners – it will take another query to get it for S3 files but I think it is possible.

**Comment by jjoyce0510** [comment]
LGTM. Thanks!

**Review by hsheth2 — APPROVED** [review]

**Review by jjoyce0510 — APPROVED** [review]
Looks good. Thanks for adding good comments + dedupe logic!

**Review by shirshanka — APPROVED** [review]
Nice work! LGTM!
